infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/AddNodeWorkflow.java

```diff
@@ -0,0 +1,205 @@
+ *
+ * @author Maithem
+ */
+@NotThreadSafe
+@Slf4j
+public class AddNodeWorkflow implements Workflow {
+
+    final AddNodeRequest request;
+
+    private Layout newLayout;
+
+    public AddNodeWorkflow(Request request) {
+        this.request = (AddNodeRequest) request;
+    }
+
+    @Override
+    public String getName() {
+        return ADD_NODE.toString();
+    }
+
+    @Override
+    public List<Action> getActions() {
+        return Arrays.asList(new BootstrapNode(),
+                new AddNodeToLayout(),
+                new StateTransfer(),
+                new MergeSegments());
+    }
+
+    class BootstrapNode implements Action {
+        @Override
+        public String getName() {
+            return "BootstrapNode";
+        }
+
+        @Override
+        public void execute(@Nonnull CorfuRuntime runtime) throws Exception {
+            try {
+                runtime.getLayoutManagementView().bootstrapNewNode(request.getEndpoint());
+            } catch (Exception e) {
+                if (e.getCause() instanceof AlreadyBootstrappedException) {
+                    log.info("BootstrapNode: Node {} already bootstrapped, skipping.", request.getEndpoint());
+                } else {
+                    throw e;
+                }
+            }
+        }
+    }
+
+
+    /**
+     * This action adds a new node to the layout. If it is also
+     * added as a logunit server, then in addition to adding
+     * the node the address space segment is split at the
+     * tail determined during the layout modification.
+     */
+    class AddNodeToLayout implements Action {
+        @Override
+        public String getName() {
+            return "AddNodeToLayout";
+        }
+
+        @Override
+        public void execute(@Nonnull CorfuRuntime runtime) throws Exception {
+
+            Layout currentLayout = (Layout) runtime.getLayoutView().getLayout().clone();
+
+            if (currentLayout.containsEndpoint(request.getEndpoint())) {
+                log.info("Node {} already exists in the layout, skipping.", request.getEndpoint());
+                newLayout = currentLayout;
+                return;
+            }
+
+            runtime.getLayoutManagementView().addNode(currentLayout, request.getEndpoint(),
+                    true, true,
+                    true, false,
+                    0);
+
+            runtime.invalidateLayout();
+            newLayout = (Layout) runtime.getLayoutView().getLayout().clone();
+            return;
+
+        }
+    }
+
+    /**
+     * Transfer an address segment from a cluster to a new node. The epoch shouldn't change
+     * during the segment transfer.
+     * @param endpoint destination node
+     * @param runtime The runtime to read the segment from
+     * @param segment segment to transfer
+     */
+    public void stateTransfer(String endpoint, CorfuRuntime runtime,
+                              Layout.LayoutSegment segment) throws Exception {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)